### PR TITLE
✨Add ManualTask Interfaces

### PR DESCRIPTION
**Pull request by alexanderkasten**
**Changes:**

1. Add ManualTask DataModel
2. Add ManualTask callback types
3. Add ManualTask Rest,Socket and EventAggregator Settings
4. Add SystemEvents for ManualTask
5. Add IManagement Interfaces for ManualTask

**Issues:**

Part of https://github.com/process-engine/process_engine_runtime/issues/91

PR: #27

---

See the [examples](#example) for inspiration.

## How can others test the changes?

> Describe how others can test your changes (you can remove this section for typo fixes etc.)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/data_models/messages/event_aggregator_settings.ts [hunk 1]
```diff
@@ -2,26 +2,33 @@
...
   signalReference: ':signal_ref',
 };
+
+export const routePaths = {
+  finishUserTask: '/processengine/correlation/:correlation_id/processinstance/:process_instance_id/node/:user_task_id/finish',
+  userTaskFinished: '/processengine/correlation/:correlation_id/processinstance/:process_instance_id/node/:user_task_id/finished',
+  finishManualTask:
+  `/processengine/correlation/${routeParams.correlationId}/processmodel/${routeParams.processModelId}/node/${routeParams.manualTaskId}/finish`,
+  manualTaskFinished:
+    `/processengine/correlation/${routeParams.correlationId}/processmodel/${routeParams.processModelId}/node/${routeParams.manualTaskId}/finished`,
```
Comment (ElRaptorus) [inline src/data_models/messages/event_aggregator_settings.ts]: Same thing as here please:
https://github.com/process-engine/consumer_api_contracts/pull/45#discussion_r232720985

File: src/data_models/messages/event_aggregator_settings.ts [hunk 2]
```diff
@@ -2,26 +2,33 @@
...
   correlationId: ':correlation_id',
   userTaskId: ':user_task_id',
+  manualTaskId: ':manual_task_id',
   messageReference: ':message_ref',
   signalReference: ':signal_ref',
 };
+
+export const routePaths = {
+  finishUserTask: '/processengine/correlation/:correlation_id/processinstance/:process_instance_id/node/:user_task_id/finish',
+  userTaskFinished: '/processengine/correlation/:correlation_id/processinstance/:process_instance_id/node/:user_task_id/finished',
```
Comment (ElRaptorus) [inline src/data_models/messages/event_aggregator_settings.ts]: Wtf ... how the hell did that get approved? 😖 

Please do me a favor and change these route paths to the same kind of templating as in the consumer API. :3

File: src/imanagement_api.ts [hunk 3]
```diff
@@ -336,6 +406,26 @@ export interface IManagementApi {
    */
   onUserTaskFinished(callback: Messages.CallbackTypes.OnUserTaskFinishedCallback): void;
 
+    /**
+   * Executes a callback when a user task is reached.
```
Comment (ElRaptorus) [inline src/imanagement_api.ts]: `ManualTask`

File: src/imanagement_api.ts [hunk 4]
```diff
@@ -336,6 +406,26 @@ export interface IManagementApi {
...
+   *
+   * @async
+   * @param callback       The callback that will be executed when a user task
```
Comment (ElRaptorus) [inline src/imanagement_api.ts]: ManualTask

File: src/imanagement_api.ts [hunk 5]
```diff
@@ -336,6 +406,26 @@ export interface IManagementApi {
...
+   *                       is reached. The message passed to the callback
+   *                       contains further information about the user task.
```
Comment (ElRaptorus) [inline src/imanagement_api.ts]: ManualTask

File: src/data_models/messages/callback_types.ts [hunk 6]
```diff
@@ -1,7 +1,17 @@
...
+  ProcessEndedMessage,
+  ProcessTerminatedMessage,
+  UserTaskFinishedMessage,
+  UserTaskReachedMessage,
+} from './system_events';
 
 export type OnUserTaskWaitingCallback = (userTaskWaiting: UserTaskReachedMessage) => void|Promise<void>;
 export type OnUserTaskFinishedCallback = (userTaskFinished: UserTaskFinishedMessage) => void|Promise<void>;
 
+export type OnManualTaskWaitingCallback = (manualTaskWaiting: ManualTaskReachedMessage) => void|Promise<void>;
```
Comment (robpal94) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnManualTaskWaitingCallback = (manualTaskWaiting: ManualTaskReachedMessage) => void | Promise<void>;
```

File: src/data_models/messages/callback_types.ts [hunk 7]
```diff
@@ -1,7 +1,17 @@
...
+export type OnManualTaskFinishedCallback = (manualTaskFinished: ManualTaskFinishedMessage) => void|Promise<void>;
```
Comment (robpal94) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnManualTaskFinishedCallback = (manualTaskFinished: ManualTaskFinishedMessage) => void | Promise<void>;
```

File: src/data_models/manual_task/manual_task.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Describes a maual task that belongs to an active correlation.
+ */
+export class ManualTask {
+  /**
+   * The id of the ManualTask.
```
Comment (cmg-dev) [inline src/data_models/manual_task/manual_task.ts]:
```suggestion
   * The ID of the ManualTask.
```

File: src/data_models/manual_task/manual_task.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+   */
+  public id: string;
+  /**
+   * The name of the ManualTask.
+   */
+  public name: string;
+  /**
+   * The id of the correlation that the ManualTask belongs to.
```
Comment (cmg-dev) [inline src/data_models/manual_task/manual_task.ts]:
```suggestion
   * The ID of the correlation that the ManualTask belongs to.
```

File: src/data_models/manual_task/manual_task.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+   */
+  public correlationId: string;
+  /**
+   * The id of the process model that the ManualTask belongs to.
```
Comment (cmg-dev) [inline src/data_models/manual_task/manual_task.ts]:
```suggestion
   * The ID of the process model that the ManualTask belongs to.
```

File: src/data_models/manual_task/manual_task.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+   */
+  public processModelId: string;
+  /**
+   * The id of the process instance that the ManualTask belongs to.
```
Comment (cmg-dev) [inline src/data_models/manual_task/manual_task.ts]:
```suggestion
   * The ID of the process instance that the ManualTask belongs to.
```

File: src/data_models/manual_task/manual_task_list.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+/**
+ * Describes a list of manual tasks.
+ */
+export class ManualTaskList {
+  /**
+   * The manual task list.
+   */
+  public manualTasks: Array<ManualTask> = [];
+}
```
Comment (cmg-dev) [inline src/data_models/manual_task/manual_task_list.ts]: Those comments are good examples of what is unnecessary as comment. you could remove them.

File: src/data_models/messages/callback_types.ts [hunk 13]
```diff
@@ -1,7 +1,17 @@
...
+  ManualTaskFinishedMessage,
+  ManualTaskReachedMessage,
+  ProcessEndedMessage,
+  ProcessTerminatedMessage,
+  UserTaskFinishedMessage,
+  UserTaskReachedMessage,
+} from './system_events';
 
 export type OnUserTaskWaitingCallback = (userTaskWaiting: UserTaskReachedMessage) => void|Promise<void>;
 export type OnUserTaskFinishedCallback = (userTaskFinished: UserTaskFinishedMessage) => void|Promise<void>;
```
Comment (cmg-dev) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnUserTaskFinishedCallback = (userTaskFinished: UserTaskFinishedMessage) => void | Promise<void>;
```

File: src/data_models/messages/callback_types.ts [hunk 14]
```diff
@@ -1,7 +1,17 @@
...
+import {
+  ManualTaskFinishedMessage,
+  ManualTaskReachedMessage,
+  ProcessEndedMessage,
+  ProcessTerminatedMessage,
+  UserTaskFinishedMessage,
+  UserTaskReachedMessage,
+} from './system_events';
 
 export type OnUserTaskWaitingCallback = (userTaskWaiting: UserTaskReachedMessage) => void|Promise<void>;
```
Comment (cmg-dev) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnUserTaskWaitingCallback = (userTaskWaiting: UserTaskReachedMessage) => void | Promise<void>;
```

File: src/data_models/messages/callback_types.ts [hunk 15]
```diff
@@ -1,7 +1,17 @@
...
 export type OnUserTaskFinishedCallback = (userTaskFinished: UserTaskFinishedMessage) => void|Promise<void>;
 
+export type OnManualTaskWaitingCallback = (manualTaskWaiting: ManualTaskReachedMessage) => void|Promise<void>;
+export type OnManualTaskFinishedCallback = (manualTaskFinished: ManualTaskFinishedMessage) => void|Promise<void>;
+
 export type OnProcessEndedCallback = (processEnded: ProcessEndedMessage) => void|Promise<void>;
 export type OnProcessTerminatedCallback = (processTerminated: ProcessTerminatedMessage) => void|Promise<void>;
```
Comment (cmg-dev) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnProcessTerminatedCallback = (processTerminated: ProcessTerminatedMessage) => void | Promise<void>;
```

File: src/data_models/messages/callback_types.ts [hunk 16]
```diff
@@ -1,7 +1,17 @@
...
+  UserTaskReachedMessage,
+} from './system_events';
 
 export type OnUserTaskWaitingCallback = (userTaskWaiting: UserTaskReachedMessage) => void|Promise<void>;
 export type OnUserTaskFinishedCallback = (userTaskFinished: UserTaskFinishedMessage) => void|Promise<void>;
 
+export type OnManualTaskWaitingCallback = (manualTaskWaiting: ManualTaskReachedMessage) => void|Promise<void>;
+export type OnManualTaskFinishedCallback = (manualTaskFinished: ManualTaskFinishedMessage) => void|Promise<void>;
+
 export type OnProcessEndedCallback = (processEnded: ProcessEndedMessage) => void|Promise<void>;
```
Comment (cmg-dev) [inline src/data_models/messages/callback_types.ts]:
```suggestion
export type OnProcessEndedCallback = (processEnded: ProcessEndedMessage) => void | Promise<void>;
```

File: src/data_models/messages/system_events/finish_manual_task.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import {BaseSystemEventMessage} from './base_system_event_message';
+
+/**
+ * The message used to finish a waiting manual task.
```
Comment (cmg-dev) [inline src/data_models/messages/system_events/finish_manual_task.ts]:
```suggestion
 * This message is used to finish a ManualTask; this assumes, that the ManualTask has been processed by the user.
 * You can put a result of the ManualTask into the `currentToken` parameter.
```

File: src/data_models/messages/system_events/finish_manual_task.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+ */
+export class FinishManualTaskMessage extends BaseSystemEventMessage {
+
+  /**
+   * The flow node id of the manual task being finished.
```
Comment (cmg-dev) [inline src/data_models/messages/system_events/finish_manual_task.ts]:
```suggestion
   * The flow node ID of the manual task being finished.
```

File: src/data_models/messages/system_events/manual_task_finished.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
+import {BaseSystemEventMessage} from './base_system_event_message';
+
+/**
+ * The message sent when a manual task has been finished.
```
Comment (cmg-dev) [inline src/data_models/messages/system_events/manual_task_finished.ts]:
```suggestion
 * The message sent when a ManualTask has been finished.
```

File: src/imanagement_api.ts [hunk 20]
```diff
@@ -336,6 +406,26 @@ export interface IManagementApi {
    */
   onUserTaskFinished(callback: Messages.CallbackTypes.OnUserTaskFinishedCallback): void;
 
+    /**
```
Comment (cmg-dev) [inline src/imanagement_api.ts]:
```suggestion
  /**
```

**Pull request merged**